### ISSUE TITLE
Slow performance for `topk`

### ISSUE THREAD

**Issue by stbaione** [iree-org/iree#20772]
### What happened?

We're working on enabling the use of `gpu kernels` in the Shortfin LLM Server for our token selection/sampling logic. This will reduce the amount of data we need to transfer from device to host and the amount of computation that we need to do on host.

However, we're seeing poor performance when exporting/compiling `topk`. The benchmark time is showing that it takes millions of milliseconds for a single execution of the `mistral` logits, w/ `k==4`.

```text
--------------------------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------
BM_topk_k4/process_time/real_time    1471283 ms      1489040 ms            1 items_per_second=679.679u/s
```

In this case, I'm using an input representative of the `mistral` logits: `<1x1x131072, f16>` and invoking a `topk` function, with `k==4`, in this case. MLIR for function is attached below.

### Steps to reproduce your issue

1. Save the following IR as `topk_k4.mlir`:
```mlir
module @module {
  func.func @topk_k4(%arg0: !torch.vtensor<[1,1,131072],f16>) -> (!torch.vtensor<[1,1,4],f16>, !torch.vtensor<[1,1,4],si64>) attributes {torch.assume_strict_symbolic_shapes} {
    %int4 = torch.constant.int 4
    %int-1 = torch.constant.int -1
    %true = torch.constant.bool true
    %true_0 = torch.constant.bool true
    %values, %indices = torch.aten.topk %arg0, %int4, %int-1, %true, %true_0 : !torch.vtensor<[1,1,131072],f16>, !torch.int, !torch.int, !torch.bool, !torch.bool -> !torch.vtensor<[1,1,4],f16>, !torch.vtensor<[1,1,4],si64>
    return %values, %indices : !torch.vtensor<[1,1,4],f16>, !torch.vtensor<[1,1,4],si64>
  }
}
```
2. Compile:
```bash
iree-compile $ARTIFACTS_DIR/mistral_$SUFFIX.mlir \
    -o $ARTIFACTS_DIR/mistral_$SUFFIX.vmfb \
    --iree-hal-target-device=hip \
    --iree-hip-target=gfx942 \
    --iree-opt-level=O3  \
    --iree-hal-indirect-command-buffers=true  \
    --iree-stream-resource-memory-model=discrete  \
    --iree-hal-memoization=true
```
3. Invoke with `iree-benchmark-module`:
```bash
iree-benchmark-module --device=hip://0 --module=mistral_artifacts/mistral_5_9_topk.vmfb --function=topk_k4 --input=1x1x131072xf16
```
4. You should see results similar to the following:
```text
--------------------------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------
BM_topk_k4/process_time/real_time    1471283 ms      1489040 ms            1 items_per_second=679.679u/s
```
4. You will probably want to change the input tensor size to a lower value and dev/test against that.

### What component(s) does this issue relate to?

Runtime

### Version information

ec7d2e995f552fc52cf360d3b1e15534e044fd4e

**Comment by MaheshRavishankar** [comment]
Can you try adding `--iree-dispatch-creation-topk-split-reduction=128` ?

**Comment by stbaione** [comment]
@MaheshRavishankar Adding that arg didn't seem to fix the issue:

```text
--------------------------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------
BM_topk_k4/process_time/real_time    1460585 ms      1485796 ms            1 items_per_second=684.657u/s
```

**Comment by hanhanW** [comment]
> Can you try adding `--iree-dispatch-creation-topk-split-reduction=128` ?

It won't help because the torch top_k is not lowered to linalg_ext.topk. It becomes a torch.aten.sort, and then becomes a linalg_ext.sort op later on.

```mlir
// -----// IR Dump After ConvertCustomQuantOp (torch-convert-custom-quant-op) //----- //
func.func @topk_k4(%arg0: !torch.vtensor<[1,1,131072],f16>) -> (!torch.vtensor<[1,1,4],f16>, !torch.vtensor<[1,1,4],si64>) attributes {torch.assume_strict_symbolic_shapes} {
  %int4 = torch.constant.int 4
  %int-1 = torch.constant.int -1
  %true = torch.constant.bool true
  %values, %indices = torch.aten.topk %arg0, %int4, %int-1, %true, %true : !torch.vtensor<[1,1,131072],f16>, !torch.int, !torch.int, !torch.bool, !torch.bool -> !torch.vtensor<[1,1,4],f16>, !torch.vtensor<[1,1,4],si64>
  return %values, %indices : !torch.vtensor<[1,1,4],f16>, !torch.vtensor<[1,1,4],si64>
}

// -----// IR Dump After DecomposeComplexOps (torch-decompose-complex-ops) //----- //
func.func @topk_k4(%arg0: !torch.vtensor<[1,1,131072],f16>) -> (!torch.vtensor<[1,1,4],f16>, !torch.vtensor<[1,1,4],si64>) attributes {torch.assume_strict_symbolic_shapes} {
  %int1 = torch.constant.int 1
  %int0 = torch.constant.int 0
  %int4 = torch.constant.int 4
  %int-1 = torch.constant.int -1
  %true = torch.constant.bool true
  %values, %indices = torch.aten.sort %arg0, %int-1, %true : !torch.vtensor<[1,1,131072],f16>, !torch.int, !torch.bool -> !torch.vtensor<[1,1,131072],f16>, !torch.vtensor<[1,1,131072],si64>
  %0 = torch.aten.slice.Tensor %values, %int-1, %int0, %int4, %int1 : !torch.vtensor<[1,1,131072],f16>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,1,4],f16>
  %1 = torch.aten.slice.Tensor %indices, %int-1, %int0, %int4, %int1 : !torch.vtensor<[1,1,131072],si64>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,1,4],si64>
  return %0, %1 : !torch.vtensor<[1,1,4],f16>, !torch.vtensor<[1,1,4],si64>
}

// -----// IR Dump After FuseQuantizedOps (torch-fuse-quantized-ops) //----- //
func.func @topk_k4(%arg0: !torch.vtensor<[1,1,131072],f16>) -> (!torch.vtensor<[1,1,4],f16>, !torch.vtensor<[1,1,4],si64>) attributes {torch.assume_strict_symbolic_shapes} {
  %int1 = torch.constant.int 1
  %int0 = torch.constant.int 0
  %int4 = torch.constant.int 4
  %int-1 = torch.constant.int -1
  %true = torch.constant.bool true
  %values, %indices = torch.aten.sort %arg0, %int-1, %true : !torch.vtensor<[1,1,131072],f16>, !torch.int, !torch.bool -> !torch.vtensor<[1,1,131072],f16>, !torch.vtensor<[1,1,131072],si64>
  %0 = torch.aten.slice.Tensor %values, %int-1, %int0, %int4, %int1 : !torch.vtensor<[1,1,131072],f16>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,1,4],f16>
  %1 = torch.aten.slice.Tensor %indices, %int-1, %int0, %int4, %int1 : !torch.vtensor<[1,1,131072],si64>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,1,4],si64>
  return %0, %1 : !torch.vtensor<[1,1,4],f16>, !torch.vtensor<[1,1,4],si64>
}

...

// -----// IR Dump After ConvertTMTensorToLinalgExtPass (torch-iree-tm-tensor-to-linalg-ext) //----- //
func.func @topk_k4(%arg0: !torch.vtensor<[1,1,131072],f16>) -> (!torch.vtensor<[1,1,4],f16>, !torch.vtensor<[1,1,4],si64>) attributes {torch.assume_strict_symbolic_shapes} {
  %int-1 = torch.constant.int -1
  %int4 = torch.constant.int 4
  %int0 = torch.constant.int 0
  %int1 = torch.constant.int 1
  %0 = torch_c.to_builtin_tensor %arg0 : !torch.vtensor<[1,1,131072],f16> -> tensor<1x1x131072xf16>
  %1 = tensor.empty() : tensor<1x1x131072xi64>
  %2 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} outs(%1 : tensor<1x1x131072xi64>) {
  ^bb0(%out: i64):
    %8 = linalg.index 2 : index
    %9 = arith.index_cast %8 : index to i64
    linalg.yield %9 : i64
  } -> tensor<1x1x131072xi64>
  %3:2 = iree_linalg_ext.sort dimension(2) outs(%0, %2 : tensor<1x1x131072xf16>, tensor<1x1x131072xi64>) {
  ^bb0(%arg1: f16, %arg2: f16, %arg3: i64, %arg4: i64):
    %8 = arith.cmpf oge, %arg1, %arg2 : f16
    iree_linalg_ext.yield %8 : i1
  } -> tensor<1x1x131072xf16>, tensor<1x1x131072xi64>
  %4 = torch_c.from_builtin_tensor %3#1 : tensor<1x1x131072xi64> -> !torch.vtensor<[1,1,131072],si64>
  %5 = torch_c.from_builtin_tensor %3#0 : tensor<1x1x131072xf16> -> !torch.vtensor<[1,1,131072],f16>
  %6 = torch.aten.slice.Tensor %5, %int-1, %int0, %int4, %int1 : !torch.vtensor<[1,1,131072],f16>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,1,4],f16>
  %7 = torch.aten.slice.Tensor %4, %int-1, %int0, %int4, %int1 : !torch.vtensor<[1,1,131072],si64>, !torch.int, !torch.int, !torch.int, !torch.int -> !torch.vtensor<[1,1,4],si64>
  return %6, %7 : !torch.vtensor<[1,1,4],f16>, !torch.vtensor<[1,1,4],si64>
}
```

Maybe @stbaione can try to not decompose the op and convert it to linalg_ext.top_k directly. It can be an option. Otherwise, using the flag does not help.

**Comment by hanhanW** [comment]
Here is an example about how it looks like in linalg_ext version:

https://github.com/iree-org/iree/blob/1d023508ce4749b468f9de2de78e53b4baffc316/tests/e2e/linalg_ext_ops/top-k.mlir#L1-L31

**Comment by MaheshRavishankar** [comment]
Hmmm, I suspected that. Sorting to do topk is strictly worse
